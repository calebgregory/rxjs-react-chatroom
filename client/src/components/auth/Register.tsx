import * as R from 'ramda'
import React, { CSSProperties } from 'react'
import { useObservableState } from 'observable-hooks'
import { pluckEventTargetValue } from 'src/observables/helpers'
import { signUp } from 'src/services/cognito-auth'

interface Props {}

export const HAS_DIGIT = 'has digit'
export const HAS_LOWER = 'has lowercase'
export const HAS_UPPER = 'has uppercase'
export const IS_ENOUGH = 'is long enough'

const validations: Map<string, RegExp> = new Map([
  [HAS_DIGIT, /[0-9]+/g],
  [HAS_LOWER, /[a-z]+/g],
  [HAS_UPPER, /[A-Z]+/g],
  [IS_ENOUGH, /^.{12,}/g],
])

export function validQualitiesFromPassword(password: string): Set<string> {
  const qualities = Array.from(validations.entries())
    .filter(([_, regex]) => R.test(regex)(password))
    .map(([quality]) => quality)

  return new Set(qualities)
}

export function getQualitiesStyle(hasQuality: boolean): CSSProperties {
  return {
    fontStyle: 'bold',
    color: hasQuality ? 'green' : 'default',
  }
}

export function getIsFormValid(validQualities: Set<string>, password: string, confirmPassword: string): boolean {
  return validQualities.size === validations.size && password === confirmPassword
}

export function Register(_: Props) {
  const [username, onUsernameChange] = useObservableState(pluckEventTargetValue(), '')
  const [email, onEmailChange] = useObservableState(pluckEventTargetValue(), '')
  const [password, onPasswordChange] = useObservableState(pluckEventTargetValue(), '')
  const [confirmPassword, onConfirmPasswordChange] = useObservableState(pluckEventTargetValue(), '')

  const validQualities = validQualitiesFromPassword(password)
  const isFormValid = getIsFormValid(validQualities, password, confirmPassword)

  const handleSubmit = (evt: React.SyntheticEvent) => {
    evt.preventDefault() // prevent refresh
    if (!isFormValid) {
      return
    }

    signUp({ username, password, email })
  }

  return (
    <form onSubmit={handleSubmit}>
      <div>
        <label>Username</label><br/>
        <input
          type="text"
          placeholder="username"
          value={username}
          onChange={onUsernameChange}
        />
      </div>
      <div>
        <label>Email</label><br />
        <input
          type="text"
          placeholder="email"
          value={email}
          onChange={onEmailChange}
        />
      </div>
      <div>
        <label>Password: must be{' '}
          <span style={getQualitiesStyle(validQualities.has(IS_ENOUGH))}>&ge; 12 characters</span>,{' '}
          include{' '}
          <span style={getQualitiesStyle(validQualities.has(HAS_UPPER))}>1. capital letter</span>,{' '}
          <span style={getQualitiesStyle(validQualities.has(HAS_LOWER))}>2. lowercase letter</span>,{' '}
          <span style={getQualitiesStyle(validQualities.has(HAS_DIGIT))}>3. a number</span>
        </label><br/>
        <input
          type="password"
          placeholder="password"
          value={password}
          onChange={onPasswordChange}
        />
      </div>
      <div>
        <label>Confirm Password</label><br/>
        <input
          type="password"
          placeholder="confirm password"
          value={confirmPassword}
          onChange={onConfirmPasswordChange}
        />
      </div>
      <div>
        <button onClick={handleSubmit} disabled={!isFormValid}>
          Register
        </button>
      </div>
    </form>
  )
}