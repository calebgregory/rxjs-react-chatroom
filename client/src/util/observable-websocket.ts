import { fromEvent, Observable, Subject, BehaviorSubject } from 'rxjs'
import { tap, share, map } from 'rxjs/operators'
import { bufferWhen } from '../observables/operators'
import { logger } from '../logger'

const log = logger('services/websocket')

export type MessageType = 'OPEN' | 'MESSAGE' | 'CLOSE' | 'ERROR'
type ConnectionStatus = 'OPEN' | 'CLOSED'

export function createOutgoing$(msg$: Observable<any>, connectionStatus$: Observable<ConnectionStatus>): Observable<any> {
  const status$ = connectionStatus$.pipe(share())
  const shouldBuffer$ = status$.pipe(map(status => status === 'CLOSED'))
  return msg$
    .pipe(
      bufferWhen(shouldBuffer$)
    )
}

export class ObservableWebSocket {
  ws: WebSocket
  incoming$: Observable<MessageEvent>
  outgoing$: Subject<any> = new Subject()
  connectionStatus$: BehaviorSubject<ConnectionStatus> = new BehaviorSubject<ConnectionStatus>('CLOSED')

  constructor(url: string) {
    this.ws = new WebSocket(url)
    this.incoming$ = fromEvent<MessageEvent>(this.ws, 'message')
      .pipe(
        tap((event: any) => { log.info('incoming msg', { event, msg: event.data }) })
      )

    createOutgoing$(this.outgoing$, this.connectionStatus$).subscribe((data: any) => {
      const msg = JSON.stringify(data)
      this.ws.send(msg)
    })

    // @TODO: buffer outgoing messages until connection is open
    this.ws.onopen = (evt: Event) => {
      log.info('ws open', { evt })
      this.connectionStatus$.next('OPEN')
    }

    this.ws.onclose = (evt: Event) => {
      log.info('ws close', { url, evt })
      this.connectionStatus$.next('CLOSED')
    }

    this.ws.onerror = (evt: Event) => {
      log.error('ws error', { url, evt })
    }
  }

  destroy() {
    // clean up
  }

  close() {
    this.ws.close(1000, 'Done')
    this.destroy()
  }

  send(data: any) {
    this.outgoing$.next(data)
  }
}