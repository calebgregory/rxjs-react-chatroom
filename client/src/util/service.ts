import { Subscription } from 'rxjs'

// @@TODO: this interface def may be superfluous
export interface Servicer {
  children: Set<Servicer>
  $subscriptions: Set<Subscription>
  destroy(): void
  $(subscription: Subscription): Subscription
  supervise(servicer: Servicer): Servicer
}

export class Service implements Servicer {
  children: Set<Service> = new Set()

  $subscriptions: Set<Subscription> = new Set()

  destroy() {
    this.$subscriptions.forEach((subscription) => { subscription.unsubscribe() })

    this.children.forEach((child) => {
      child.destroy()
    })
  }

  $(subscription: Subscription): Subscription {
    this.$subscriptions.add(subscription)
    return subscription
  }

  supervise<T>(child: unknown) { // @@TODO: fix this typing
    this.children.add(child as Service)
    return child as T
  }
}