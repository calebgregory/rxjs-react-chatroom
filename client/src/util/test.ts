import * as R from 'ramda'
import { TestScheduler } from 'rxjs/testing'

export function makeTestScheduler() {
  const testScheduler = new TestScheduler((actual, expected) => {
    expect(actual).toEqual(expected)
  })
  return testScheduler
}

/*
{1,2,3}

{1,2,3} 1 {2,3} 2 {3} 3 {}
{1,2,}  1 {2}
{1,3,}  1 {3}
{2,3,}
{1,}
{2,}
{3,}
{}
*/

export function powerset(S: Set<any>): Set<Set<any>> {
  let i = 0
  const _ps = (list: Array<any>): Array<Array<any>> => {
    if (list.length === 0) {
      return [[]] // base case 'empty set'
    }

    const [ head, ...tail ] = list // pick first
    const tailPS = _ps(tail)       // recur rest
    const subsets = [              // combine
      ...tailPS,                   //             the other subsets
      ...tailPS.map((subset) => [head, ...subset]) // with each of those subsets' elements, combined with the one we picked
    ] // ,-o-o if you've picked one out of a set, it will not be in the the remaining set's subsets

    return subsets
  }
  const subsets: Array<Array<any>> = _ps(Array.from(S))
  return new Set(subsets.map((subset) => new Set(subset)))
}