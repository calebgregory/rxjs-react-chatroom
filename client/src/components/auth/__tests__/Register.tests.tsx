import { powerset } from 'src/util/test'
import * as c from '../Register'

describe('validQualitiesFromPassword', () => {
  it('says when password has lowercase letter', () => {
    const validQualities = c.validQualitiesFromPassword('a')
    expect(validQualities).toEqual(new Set([c.HAS_LOWER]))
  })

  it('says when password has uppercase letter', () => {
    const validQualities = c.validQualitiesFromPassword('A')
    expect(validQualities).toEqual(new Set([c.HAS_UPPER]))
  })

  it('says when password has digit', () => {
    const validQualities = c.validQualitiesFromPassword('1')
    expect(validQualities).toEqual(new Set([c.HAS_DIGIT]))
  })


  it('says when password is long enough', () => {
    const validQualities = c.validQualitiesFromPassword('asdfas09fja0sdfasdf')
    expect(validQualities.has(c.IS_ENOUGH)).toEqual(true)
  })
})

describe('getQualitiesStyle', () => {
  it('has color = green when hasQuality = true', () => {
    const styles = c.getQualitiesStyle(true)
    expect(styles).toMatchSnapshot()
  })
})

describe('getIsFormValid', () => {
  const allQualities = new Set([c.HAS_DIGIT, c.HAS_LOWER, c.HAS_UPPER, c.IS_ENOUGH])
  const qualitiesPowerset: Set<Set<string>> = powerset(allQualities)
  const strictSubset = Array.from(qualitiesPowerset).filter((subset) => subset === allQualities) // is actually strict subset now...

  test.each([
    [allQualities, 'a', 'a', true],
    [allQualities, 'aa', 'a', false],
    ...strictSubset.map((subset: Set<string>) => [subset, 'a', 'a', false])
  ])('%p %s %s -> %p', (validQualities, password, confirmPassword, expected) => {
    const isValid = c.getIsFormValid(validQualities as Set<string>, password as string, confirmPassword as string)
    expect(isValid).toEqual(expected)
  })
})