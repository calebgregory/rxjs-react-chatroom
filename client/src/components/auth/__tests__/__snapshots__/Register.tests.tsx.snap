// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getQualitiesStyle has color = green when hasQuality = true 1`] = `
Object {
  "color": "green",
  "fontStyle": "bold",
}
`;
